Questions.

1. How the number of connectors affect the charge time
2. How the number of connectors affect charge KWh
3. How the model affect the kWh
4. How the modell affect the Charge time
5. The individual differences between sites on charge time
df.duplicated().sum() # There are no duplicate rows

In [24]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np


# Permanently changes the pandas settings
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)
pd.set_option('display.width', None)
pd.set_option('display.max_colwidth', -1)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:10: FutureWarning: Passing a negative integer is deprecated in version 1.0 and will not be supported in future version. Instead, use None to not limit the column width.
  # Remove the CWD from sys.path while we load stuff.


In [25]:
pd.reset_option('all')

As the xlwt package is no longer maintained, the xlwt engine will be removed in a future version of pandas. This is the only engine in pandas that supports writing in the xls format. Install openpyxl and write to an xlsx file instead.

: boolean
    use_inf_as_null had been deprecated and will be removed in a future
    version. Use `use_inf_as_na` instead.



/usr/local/lib/python3.7/dist-packages/pandas/_config/config.py:630: FutureWarning: As the xlwt package is no longer maintained, the xlwt engine will be removed in a future version of pandas. This is the only engine in pandas that supports writing in the xls format. Install openpyxl and write to an xlsx file instead.
  warnings.warn(d.msg, FutureWarning)
/usr/local/lib/python3.7/dist-packages/pandas/_config/config.py:630: FutureWarning: 
: boolean
    use_inf_as_null had been deprecated and will be removed in a future
    version. Use `use_inf_as_na` instead.

  warnings.warn(d.msg, FutureWarning)


In [26]:

df = pd.read_csv('/content/2019 EVCP use Q3 and Q4.csv')
df

,Charging event,User ID,CP ID,Connector,Start Date,Start Time,End Date,End Time,Total kWh,Site,Model
0,8703401,User 315,70206,2,07/01/2019,07:44,NaN,NaN,NaN,Elland Road Park and Ride,APT 7kW Dual Outlet
1,8703471,User 412,80080,1,07/01/2019,07:59,07/01/2019,17:31,6.55,Temple Green Park and Ride,APT 7kW Dual Outlet
2,8703492,User 546,70204,2,07/01/2019,08:05,07/01/2019,17:13,9.01,Woodhouse Lane Car Park,APT 7kW Dual Outlet
3,8703730,User 379,80085,2,07/01/2019,08:50,07/01/2019,18:02,4.22,Temple Green Park and Ride,APT 7kW Dual Outlet
4,8703829,User 712,70203,1,07/01/2019,09:07,07/01/2019,18:01,27.48,Woodhouse Lane Car Park,APT 7kW Dual Outlet
...,...,...,...,...,...,...,...,...,...,...,...
4210,9397977,User 315,70206,2,12/31/2019,07:11,12/31/2019,13:20,12.91,Elland Road Park and Ride,APT 7kW Dual Outlet
4211,9397985,User 546,70204,1,12/31/2019,07:14,12/31/2019,13:42,8.59,Woodhouse Lane Car Park,APT 7kW Dual Outlet
4212,9398179,User 388,70203,1,12/31/2019,08:15,12/31/2019,14:05,13.06,Woodhouse Lane Car Park,APT 7kW Dual Outlet
4213,9398486,User 494,70202,2,12/31/2019,09:27,12/31/2019,19:03,32.11,Woodhouse Lane Car Park,APT 7kW Dual Outlet


The data in question has 4215 observations and 11 attributes.

Charging event has no empty data and it is numeric with a minimum of 8703401 and a maximum of 9399679. It is the serial count of the events

User ID is the unique ID for each event. The use of "User" preceeding the digit is redundant as it is repeated in every event.

CP ID 



In [ ]:
shape = df.shape
shape
df.info()
print(df.describe())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4215 entries, 0 to 4214
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Charging event  4215 non-null   int64  
 1   User ID         4215 non-null   object 
 2   CP ID           4215 non-null   int64  
 3   Connector       4215 non-null   int64  
 4   Start Date      4215 non-null   object 
 5   Start Time      4215 non-null   object 
 6   End Date        4147 non-null   object 
 7   End Time        4147 non-null   object 
 8   Total kWh       4147 non-null   float64
 9   Site            4215 non-null   object 
 10  Model           4215 non-null   object 
dtypes: float64(1), int64(3), object(7)
memory usage: 362.4+ KB
       Charging event         CP ID    Connector    Total kWh
count    4.215000e+03   4215.000000  4215.000000  4147.000000
mean     9.051927e+06  69058.089917     1.539739    11.929197
std      2.013920e+05   9262.732943     0.498477     9.243

In [27]:
df["Quater"] = df['Start Date'].str.split('/')

In [28]:
df

,Charging event,User ID,CP ID,Connector,Start Date,Start Time,End Date,End Time,Total kWh,Site,Model,Quater
0,8703401,User 315,70206,2,07/01/2019,07:44,NaN,NaN,NaN,Elland Road Park and Ride,APT 7kW Dual Outlet,"[07, 01, 2019]"
1,8703471,User 412,80080,1,07/01/2019,07:59,07/01/2019,17:31,6.55,Temple Green Park and Ride,APT 7kW Dual Outlet,"[07, 01, 2019]"
2,8703492,User 546,70204,2,07/01/2019,08:05,07/01/2019,17:13,9.01,Woodhouse Lane Car Park,APT 7kW Dual Outlet,"[07, 01, 2019]"
3,8703730,User 379,80085,2,07/01/2019,08:50,07/01/2019,18:02,4.22,Temple Green Park and Ride,APT 7kW Dual Outlet,"[07, 01, 2019]"
4,8703829,User 712,70203,1,07/01/2019,09:07,07/01/2019,18:01,27.48,Woodhouse Lane Car Park,APT 7kW Dual Outlet,"[07, 01, 2019]"
...,...,...,...,...,...,...,...,...,...,...,...,...
4210,9397977,User 315,70206,2,12/31/2019,07:11,12/31/2019,13:20,12.91,Elland Road Park and Ride,APT 7kW Dual Outlet,"[12, 31, 2019]"
4211,9397985,User 546,70204,1,12/31/2019,07:14,12/31/2019,13:42,8.59,Woodhouse Lane Car Park,APT 7kW Dual Outlet,"[12, 31, 2019]"
4212,9398179,User 388,70203,1,12/31/2019,08:15,12/31/2019,14:05,13.06,Woodhouse Lane Car Park,APT 7kW Dual Outlet,"[12, 31, 2019]"
4213,9398486,User 494,70202,2,12/31/2019,09:27,12/31/2019,19:03,32.11,Woodhouse Lane Car Park,APT 7kW Dual Outlet,"[12, 31, 2019]"


In [29]:
df["Quat"] = df['Quater'].apply(lambda x: 'Q3' if x[0] in ['07', '08', '09'] else 'Q4')
df

,Charging event,User ID,CP ID,Connector,Start Date,Start Time,End Date,End Time,Total kWh,Site,Model,Quater,Quat
0,8703401,User 315,70206,2,07/01/2019,07:44,NaN,NaN,NaN,Elland Road Park and Ride,APT 7kW Dual Outlet,"[07, 01, 2019]",Q3
1,8703471,User 412,80080,1,07/01/2019,07:59,07/01/2019,17:31,6.55,Temple Green Park and Ride,APT 7kW Dual Outlet,"[07, 01, 2019]",Q3
2,8703492,User 546,70204,2,07/01/2019,08:05,07/01/2019,17:13,9.01,Woodhouse Lane Car Park,APT 7kW Dual Outlet,"[07, 01, 2019]",Q3
3,8703730,User 379,80085,2,07/01/2019,08:50,07/01/2019,18:02,4.22,Temple Green Park and Ride,APT 7kW Dual Outlet,"[07, 01, 2019]",Q3
4,8703829,User 712,70203,1,07/01/2019,09:07,07/01/2019,18:01,27.48,Woodhouse Lane Car Park,APT 7kW Dual Outlet,"[07, 01, 2019]",Q3
...,...,...,...,...,...,...,...,...,...,...,...,...,...
4210,9397977,User 315,70206,2,12/31/2019,07:11,12/31/2019,13:20,12.91,Elland Road Park and Ride,APT 7kW Dual Outlet,"[12, 31, 2019]",Q4
4211,9397985,User 546,70204,1,12/31/2019,07:14,12/31/2019,13:42,8.59,Woodhouse Lane Car Park,APT 7kW Dual Outlet,"[12, 31, 2019]",Q4
4212,9398179,User 388,70203,1,12/31/2019,08:15,12/31/2019,14:05,13.06,Woodhouse Lane Car Park,APT 7kW Dual Outlet,"[12, 31, 2019]",Q4
4213,9398486,User 494,70202,2,12/31/2019,09:27,12/31/2019,19:03,32.11,Woodhouse Lane Car Park,APT 7kW Dual Outlet,"[12, 31, 2019]",Q4


In [30]:
df = df.drop(['Quater'], axis=1)
df

,Charging event,User ID,CP ID,Connector,Start Date,Start Time,End Date,End Time,Total kWh,Site,Model,Quat
0,8703401,User 315,70206,2,07/01/2019,07:44,NaN,NaN,NaN,Elland Road Park and Ride,APT 7kW Dual Outlet,Q3
1,8703471,User 412,80080,1,07/01/2019,07:59,07/01/2019,17:31,6.55,Temple Green Park and Ride,APT 7kW Dual Outlet,Q3
2,8703492,User 546,70204,2,07/01/2019,08:05,07/01/2019,17:13,9.01,Woodhouse Lane Car Park,APT 7kW Dual Outlet,Q3
3,8703730,User 379,80085,2,07/01/2019,08:50,07/01/2019,18:02,4.22,Temple Green Park and Ride,APT 7kW Dual Outlet,Q3
4,8703829,User 712,70203,1,07/01/2019,09:07,07/01/2019,18:01,27.48,Woodhouse Lane Car Park,APT 7kW Dual Outlet,Q3
...,...,...,...,...,...,...,...,...,...,...,...,...
4210,9397977,User 315,70206,2,12/31/2019,07:11,12/31/2019,13:20,12.91,Elland Road Park and Ride,APT 7kW Dual Outlet,Q4
4211,9397985,User 546,70204,1,12/31/2019,07:14,12/31/2019,13:42,8.59,Woodhouse Lane Car Park,APT 7kW Dual Outlet,Q4
4212,9398179,User 388,70203,1,12/31/2019,08:15,12/31/2019,14:05,13.06,Woodhouse Lane Car Park,APT 7kW Dual Outlet,Q4
4213,9398486,User 494,70202,2,12/31/2019,09:27,12/31/2019,19:03,32.11,Woodhouse Lane Car Park,APT 7kW Dual Outlet,Q4


In [31]:
df["Start DateTime"] = df["Start Date"] +' ' + df["Start Time"]
df

,Charging event,User ID,CP ID,Connector,Start Date,Start Time,End Date,End Time,Total kWh,Site,Model,Quat,Start DateTime
0,8703401,User 315,70206,2,07/01/2019,07:44,NaN,NaN,NaN,Elland Road Park and Ride,APT 7kW Dual Outlet,Q3,07/01/2019 07:44
1,8703471,User 412,80080,1,07/01/2019,07:59,07/01/2019,17:31,6.55,Temple Green Park and Ride,APT 7kW Dual Outlet,Q3,07/01/2019 07:59
2,8703492,User 546,70204,2,07/01/2019,08:05,07/01/2019,17:13,9.01,Woodhouse Lane Car Park,APT 7kW Dual Outlet,Q3,07/01/2019 08:05
3,8703730,User 379,80085,2,07/01/2019,08:50,07/01/2019,18:02,4.22,Temple Green Park and Ride,APT 7kW Dual Outlet,Q3,07/01/2019 08:50
4,8703829,User 712,70203,1,07/01/2019,09:07,07/01/2019,18:01,27.48,Woodhouse Lane Car Park,APT 7kW Dual Outlet,Q3,07/01/2019 09:07
...,...,...,...,...,...,...,...,...,...,...,...,...,...
4210,9397977,User 315,70206,2,12/31/2019,07:11,12/31/2019,13:20,12.91,Elland Road Park and Ride,APT 7kW Dual Outlet,Q4,12/31/2019 07:11
4211,9397985,User 546,70204,1,12/31/2019,07:14,12/31/2019,13:42,8.59,Woodhouse Lane Car Park,APT 7kW Dual Outlet,Q4,12/31/2019 07:14
4212,9398179,User 388,70203,1,12/31/2019,08:15,12/31/2019,14:05,13.06,Woodhouse Lane Car Park,APT 7kW Dual Outlet,Q4,12/31/2019 08:15
4213,9398486,User 494,70202,2,12/31/2019,09:27,12/31/2019,19:03,32.11,Woodhouse Lane Car Park,APT 7kW Dual Outlet,Q4,12/31/2019 09:27


In [34]:
pd.isna(df).sum()

#There are 68 empty cells on the following attributes End date, End Time and KWh
# Owing to the fact that it is not possible to use the input methods from statistical methods. Hence dropping the row is recommended.

Charging event     0
User ID            0
CP ID              0
Connector          0
Start Date         0
Start Time         0
End Date          68
End Time          68
Total kWh         68
Site               0
Model              0
Quat               0
Start DateTime     0
dtype: int64

In [36]:
df = df.dropna()
df.shape

(4147, 13)

In [37]:
df

,Charging event,User ID,CP ID,Connector,Start Date,Start Time,End Date,End Time,Total kWh,Site,Model,Quat,Start DateTime
1,8703471,User 412,80080,1,07/01/2019,07:59,07/01/2019,17:31,6.55,Temple Green Park and Ride,APT 7kW Dual Outlet,Q3,07/01/2019 07:59
2,8703492,User 546,70204,2,07/01/2019,08:05,07/01/2019,17:13,9.01,Woodhouse Lane Car Park,APT 7kW Dual Outlet,Q3,07/01/2019 08:05
3,8703730,User 379,80085,2,07/01/2019,08:50,07/01/2019,18:02,4.22,Temple Green Park and Ride,APT 7kW Dual Outlet,Q3,07/01/2019 08:50
4,8703829,User 712,70203,1,07/01/2019,09:07,07/01/2019,18:01,27.48,Woodhouse Lane Car Park,APT 7kW Dual Outlet,Q3,07/01/2019 09:07
5,8703850,User 279,80080,2,07/01/2019,09:11,07/01/2019,18:20,7.75,Temple Green Park and Ride,APT 7kW Dual Outlet,Q3,07/01/2019 09:11
...,...,...,...,...,...,...,...,...,...,...,...,...,...
4210,9397977,User 315,70206,2,12/31/2019,07:11,12/31/2019,13:20,12.91,Elland Road Park and Ride,APT 7kW Dual Outlet,Q4,12/31/2019 07:11
4211,9397985,User 546,70204,1,12/31/2019,07:14,12/31/2019,13:42,8.59,Woodhouse Lane Car Park,APT 7kW Dual Outlet,Q4,12/31/2019 07:14
4212,9398179,User 388,70203,1,12/31/2019,08:15,12/31/2019,14:05,13.06,Woodhouse Lane Car Park,APT 7kW Dual Outlet,Q4,12/31/2019 08:15
4213,9398486,User 494,70202,2,12/31/2019,09:27,12/31/2019,19:03,32.11,Woodhouse Lane Car Park,APT 7kW Dual Outlet,Q4,12/31/2019 09:27


In [38]:
df["End DateTime"] = df["End Date"] +' ' + df["End Time"]
df

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


,Charging event,User ID,CP ID,Connector,Start Date,Start Time,End Date,End Time,Total kWh,Site,Model,Quat,Start DateTime,End DateTime
1,8703471,User 412,80080,1,07/01/2019,07:59,07/01/2019,17:31,6.55,Temple Green Park and Ride,APT 7kW Dual Outlet,Q3,07/01/2019 07:59,07/01/2019 17:31
2,8703492,User 546,70204,2,07/01/2019,08:05,07/01/2019,17:13,9.01,Woodhouse Lane Car Park,APT 7kW Dual Outlet,Q3,07/01/2019 08:05,07/01/2019 17:13
3,8703730,User 379,80085,2,07/01/2019,08:50,07/01/2019,18:02,4.22,Temple Green Park and Ride,APT 7kW Dual Outlet,Q3,07/01/2019 08:50,07/01/2019 18:02
4,8703829,User 712,70203,1,07/01/2019,09:07,07/01/2019,18:01,27.48,Woodhouse Lane Car Park,APT 7kW Dual Outlet,Q3,07/01/2019 09:07,07/01/2019 18:01
5,8703850,User 279,80080,2,07/01/2019,09:11,07/01/2019,18:20,7.75,Temple Green Park and Ride,APT 7kW Dual Outlet,Q3,07/01/2019 09:11,07/01/2019 18:20
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
4210,9397977,User 315,70206,2,12/31/2019,07:11,12/31/2019,13:20,12.91,Elland Road Park and Ride,APT 7kW Dual Outlet,Q4,12/31/2019 07:11,12/31/2019 13:20
4211,9397985,User 546,70204,1,12/31/2019,07:14,12/31/2019,13:42,8.59,Woodhouse Lane Car Park,APT 7kW Dual Outlet,Q4,12/31/2019 07:14,12/31/2019 13:42
4212,9398179,User 388,70203,1,12/31/2019,08:15,12/31/2019,14:05,13.06,Woodhouse Lane Car Park,APT 7kW Dual Outlet,Q4,12/31/2019 08:15,12/31/2019 14:05
4213,9398486,User 494,70202,2,12/31/2019,09:27,12/31/2019,19:03,32.11,Woodhouse Lane Car Park,APT 7kW Dual Outlet,Q4,12/31/2019 09:27,12/31/2019 19:03


In [40]:
df['Start DateTime Convert'] = pd.to_datetime(df['Start DateTime'], 
                                      format="%m/%d/%Y %H:%M")

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [41]:
df

,Charging event,User ID,CP ID,Connector,Start Date,Start Time,End Date,End Time,Total kWh,Site,Model,Quat,Start DateTime,End DateTime,Start DateTime Convert
1,8703471,User 412,80080,1,07/01/2019,07:59,07/01/2019,17:31,6.55,Temple Green Park and Ride,APT 7kW Dual Outlet,Q3,07/01/2019 07:59,07/01/2019 17:31,2019-07-01 07:59:00
2,8703492,User 546,70204,2,07/01/2019,08:05,07/01/2019,17:13,9.01,Woodhouse Lane Car Park,APT 7kW Dual Outlet,Q3,07/01/2019 08:05,07/01/2019 17:13,2019-07-01 08:05:00
3,8703730,User 379,80085,2,07/01/2019,08:50,07/01/2019,18:02,4.22,Temple Green Park and Ride,APT 7kW Dual Outlet,Q3,07/01/2019 08:50,07/01/2019 18:02,2019-07-01 08:50:00
4,8703829,User 712,70203,1,07/01/2019,09:07,07/01/2019,18:01,27.48,Woodhouse Lane Car Park,APT 7kW Dual Outlet,Q3,07/01/2019 09:07,07/01/2019 18:01,2019-07-01 09:07:00
5,8703850,User 279,80080,2,07/01/2019,09:11,07/01/2019,18:20,7.75,Temple Green Park and Ride,APT 7kW Dual Outlet,Q3,07/01/2019 09:11,07/01/2019 18:20,2019-07-01 09:11:00
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
4210,9397977,User 315,70206,2,12/31/2019,07:11,12/31/2019,13:20,12.91,Elland Road Park and Ride,APT 7kW Dual Outlet,Q4,12/31/2019 07:11,12/31/2019 13:20,2019-12-31 07:11:00
4211,9397985,User 546,70204,1,12/31/2019,07:14,12/31/2019,13:42,8.59,Woodhouse Lane Car Park,APT 7kW Dual Outlet,Q4,12/31/2019 07:14,12/31/2019 13:42,2019-12-31 07:14:00
4212,9398179,User 388,70203,1,12/31/2019,08:15,12/31/2019,14:05,13.06,Woodhouse Lane Car Park,APT 7kW Dual Outlet,Q4,12/31/2019 08:15,12/31/2019 14:05,2019-12-31 08:15:00
4213,9398486,User 494,70202,2,12/31/2019,09:27,12/31/2019,19:03,32.11,Woodhouse Lane Car Park,APT 7kW Dual Outlet,Q4,12/31/2019 09:27,12/31/2019 19:03,2019-12-31 09:27:00


In [43]:
df.duplicated().sum() # There are no duplicate rows

0